src/node_dir.h

```diff
@@ -0,0 +1,87 @@
+#ifndef SRC_NODE_DIR_H_
+#define SRC_NODE_DIR_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node_file.h"
+#include "node.h"
+#include "req_wrap-inl.h"
+
+namespace node {
+
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Null;
+using v8::Object;
+using v8::Persistent;
+using v8::Promise;
+using v8::Value;
+
+using fs::FSReqBase;
+using fs::FSReqAfterScope;
+using fs::FSReqWrapSync;
+
+namespace fs_dir {
+
+class DirHandle;
+class DirReq;
+
+// XXX Needed to have access to the DirHandle in async opendir so that we can
+// set the uv_dir_t.
+class DirReq : public FSReqBase {
+ public:
+  static DirReq* from_req(uv_fs_t* req) {
+    return static_cast<DirReq*>(ReqWrap::from_req(req));
+  }
+
+  DirHandle* handle_;
```

Comment:
What happens if the handle is garbage collected while the `DirReq` operation is in progress?

Reply:
Thinkin about it, it might be easiest to store the `DirHandle` as a JS property on the `FSReqCallback` instance – that would remove the need for a subclass altogether (i.e. also address https://github.com/nodejs/node/pull/29349#discussion_r318357964), and remove any garbage collection concerns.

Reply:
No, already marked this comment and the other one as resolved :)